File: aoc2316.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {

	var A [][]string
	var line string

	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		line = scanner.Text()
		temp := make([]string, len(line))
		for i, c := range line {
			temp[i] = string(c)
		}
		A = append(A, temp)
	}
	if len(line) < 1 {
		os.Exit(0)
	}

	r1 := BFS(A, [2]int{0, -1}, [2]int{0, 1})

	var r2 int
	R := len(A)
	C := len(A[0])
	for r := 0; r < R; r++ {
		r2 = max(r2, BFS(A, [2]int{r, -1}, [2]int{0, 1}))
		r2 = max(r2, BFS(A, [2]int{r, C}, [2]int{0, -1}))
	}
	for c := 0; c < C; c++ {
		r2 = max(r2, BFS(A, [2]int{-1, c}, [2]int{1, 0}))
		r2 = max(r2, BFS(A, [2]int{R, c}, [2]int{-1, 0}))
	}
	fmt.Println("part 1:", r1)
	fmt.Println("part 2:", r2)
}

func BFS(A [][]string, curr [2]int, move [2]int) int {

	R, C := len(A), len(A[0])
	S := make(map[[2][2]int]bool)
	D := [][2][2]int{{curr, move}}

	for len(D) > 0 {

			curr, move := D[0][0], D[0][1]
		D = D[1:]
		r, c := curr[0], curr[1]
		r += move[0]
		c += move[1]

		if r < 0 || r > R-1 || c < 0 || c > C-1 {
			continue
		}

		curr = [2]int{r, c}
		t := A[r][c]
		var todo [][2][2]int

		if t == "." {

			coor := [2][2]int{curr, move}
			todo = append(todo, coor)
		} else if t == "/" {

			move = [2]int{-move[1], -move[0]}
			coor := [2][2]int{curr, move}
			todo = append(todo, coor)
		} else if t == "\\" {

			move = [2]int{move[1], move[0]}
			coor := [2][2]int{curr, move}
			todo = append(todo, coor)
		} else if t == "-" {

			if move[1] != 0 {

				coor := [2][2]int{curr, move}
				todo = append(todo, coor)
			} else { // split into 2 <--- moving horizontal

				move = [2]int{0, -1}
				coor := [2][2]int{curr, move}
				todo = append(todo, coor)
				move = [2]int{0, 1}
				coor = [2][2]int{curr, move}
				todo = append(todo, coor)
			}
		} else if t == "|" {

			if move[0] != 0 {
				coor := [2][2]int{curr, move}
				todo = append(todo, coor)
			} else {

				move = [2]int{1, 0}
				coor := [2][2]int{curr, move}
				todo = append(todo, coor)
				move = [2]int{-1, 0}
				coor = [2][2]int{curr, move}
				todo = append(todo, coor)
			}
		}

		for _, job := range todo {
			if S[job] {
				continue
			}
			D = append(D, job)
			S[job] = true
		}
	}

	SS := make(map[[2]int]bool)
	for p := range S {
		SS[p[0]] = true
	}
	return len(SS)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: aoc2316.py
F = 0
r1, r2 = 0, 0
A = []
def dbg(T):
    for t in T:print(' '.join(t))
    print()
with open('16.' + str(F)) as file:
    for line in file:
        A.append([_ for _ in line.strip()])

R, C = len(A), len(A[0])
S = set()
#print(A)
#D = [((0, -1), (0, 1))] # top left moving right
#S.add((((0, -1), (0, 1))))

def BFS(A, curr, move) -> int:

    D = [(curr, move)] # begin from top left moving right
    R, C = len(A), len(A[0])
    S = set()

    while D:
        curr, move = D.pop(0)
        r, c = curr
        r += move[0]
        c += move[1]
        if r < 0 or r > R - 1 or c < 0 or c > C - 1:
            continue
        curr = (r, c)
        t = A[r][c]
        todo = []
        if t == '.':# can skip
            coor = (curr, move)
            todo.append(coor)
        elif t == '/':
            move = (-move[1], -move[0])
            coor = (curr, move)
            todo.append(coor)
        elif t == '\\':
            move = move[::-1]
            coor = (curr, move)
            todo.append(coor)
        elif t == '|':
            if move[0] != 0: # riding
                coor = (curr, move)
                todo.append(coor)
            else:
                for rr in [-1, 1]:
                    move = (rr, 0)
                    coor = (curr, move)
                    todo.append(coor)
        elif t == '-':
            if move[1] != 0: # riding
                coor = (curr, move)
                todo.append(coor)
            else: # split into 2 <--- moving horizontal
                for cc in [-1, 1]:
                    move = (0, cc)
                    coor = (curr, move)
                    todo.append(coor)
        for job in todo:
            if job in S: continue
            D.append(job)
            S.add(job)
    SS = set()
    for p, _ in S:
        SS.add(p)
    return len(SS)

# part 1
r1 = BFS(A, (0, -1),(0,1))

# part 2
for r in range(R):
    r2 = max(r2, BFS(A, (r, -1), (0,  1)))
    r2 = max(r2, BFS(A, (r,  C), (0, -1)))
for c in range(C):
    r2 = max(r2, BFS(A, (-1, c), ( 1, 0)))
    r2 = max(r2, BFS(A, ( R, c), (-1, 0)))

print("part 1:", r1)
print("part 2:", r2)


